witness/src/Dscp/Witness/Web/Logic.hs

```diff
@@ -29,17 +29,31 @@ import Dscp.Witness.Web.Types
 -- Logic
 ----------------------------------------------------------------------------
 
+-- | Check that source account has enough money to make transaction.
+ensureEnoughMoney :: WitnessWorkMode ctx m => TxWitnessed -> m ()
+ensureEnoughMoney tw = do
+    let tx     = twTx tw
+        spent  = txInValue tx
+        source = tiaAddr (txInAcc tx)
+    account <- fromMaybe def <$> getMempoolAccountMaybe source
+
+    when (aBalance account < coinToInteger spent) $
+        throwM NotEnoughMoney
+
 -- | Applies transaction everywhere.
 -- Once call of this function returns, transaction gets considered by
 -- 'getAccountInfo' and other endpoints.
 submitUserTx :: WitnessWorkMode ctx m => TxWitnessed -> m ()
-submitUserTx tw =
+submitUserTx tw = do
+    ensureEnoughMoney tw
     Relay.relayTx (GMoneyTxWitnessed tw) >>= wait @"tx in mempool"
 
 -- | Applies transaction, but does not wait for a whole cycle of transaction
 -- application.
 submitUserTxAsync :: WitnessWorkMode ctx m => TxWitnessed -> m ()
-submitUserTxAsync = void . Relay.relayTx . GMoneyTxWitnessed
+submitUserTxAsync tw = do
+    ensureEnoughMoney tw
```

Comment:
By the way, I did it in validator.

Reply:
Yes, but as far as I understand it happens beyond relay, and transaction will just silently fail unless I check it once again here.

Reply:
They'd rather be put into `failedTxs` map.

Reply:
Would would be a proper way to do that?
AFAIU `failedTxs` is available in worker context only, should I move it to node context?

Reply:
This seems like a kostyl. I think that the proper way of doing this is to rethrow exceptions through relay somehow. Otherwise we'll have to build such kostyls for every reasonable error which might happen with a transaction (I guess "invalid signature" goes here too, i. e. now API would just silently fail).
We might leave it as is if doing things properly turns out to be really complicated, but we _need_ to do it eventually.